facilitator/src/logging.rs

```diff
@@ -1,67 +1,169 @@
         }
     }
 }
 
-#[derive(Debug, Serialize)]
-#[serde(rename_all = "camelCase")]
-struct LogEntry<'a> {
-    message: String,
-    severity: Severity,
-    time: String,
-    module_path: Option<&'a str>,
-    file: Option<&'a str>,
-    line: Option<u32>,
-    thread_id: String,
+/// Options for configuring logging in this application
+pub struct LoggingConfiguration {
+    /// If true, logging output will be forced to JSON format using
+    /// [slog-json][1]. If false, logging format will be determined by detecting
+    /// whether `stderr` is a `tty`. If it is, output is formatted using
+    /// [slog-term][2]. Otherwise, `slog-json` is used.
+    ///
+    /// [1]: https://docs.rs/slog-json
+    /// [2]: https://docs.rs/slog-term
+    pub force_json_output: bool,
+    /// A version string which shall be attached to all log messages
+    pub version_string: &'static str,
+    /// Messages above this log level will be discarded
+    pub log_level: &'static str,
 }
 
-impl<'a> LogEntry<'a> {
-    fn from_record(fmt: &Formatter, record: &'a Record) -> LogEntry<'a> {
-        LogEntry {
-            severity: Severity::from_log_level(record.level()),
-            time: fmt.timestamp().to_string(),
-            message: record.args().to_string(),
-            module_path: record.module_path(),
-            file: record.file(),
-            line: record.line(),
-            thread_id: format!("{:?}", thread::current().id()),
-        }
-    }
+/// Initialize logging resources. On success, returns a root [`slog::Logger`][1]
+/// from which child loggers may be built and a GlobalLoggerGuard which must be
+/// kept live by the caller to enable the `slog_scope` global logger to function
+/// in modules that haven't yet opted into managing their own `slog::Logger`.
+/// Returns an error if `LoggingConfiguration` is invalid.
+///
+/// [1]: https://docs.rs/slog/2.7.0/slog/struct.Logger.html
+pub fn setup_logging(config: &LoggingConfiguration) -> Result<(Logger, GlobalLoggerGuard)> {
+    // We have to box the Drain so that both branches return the same type
+    let drain: Box<dyn Drain<Ok = (), Err = std::io::Error> + Send> =
+        if atty::isnt(Stream::Stderr) || config.force_json_output {
+            // If stderr is not a tty, output logs as JSON structures on the
+            // assumption that we are running in a cloud.
+            let json_drain = Json::new(stderr())
+                .set_newlines(true)
+                // slog_json::JsonBuilder::add_default_keys adds keys for
+                // timestamp, level and the actual message, but we need the JSON
+                // keys to match what the Google Cloud Logging agent expects:
+                // https://cloud.google.com/logging/docs/agent/configuration#process-payload
+                .add_key_value(o!(
+                    "time" => PushFnValue(|_, serializer| {
+                        serializer.emit(chrono::Local::now().to_rfc3339())
+                    }),
```

Comment:
https://docs.rs/slog/2.7.0/slog/struct.PushFnValue.html says:

> It's more ergonomic for closures used as lazy values to return type implementing Serialize, but sometimes that forces an allocation (eg. Strings)

Since this is called once, I don't think we're worried about allocations. Can you use `chrono::Local::now().to_rfc3339()` directly? Presumably its output is String and String implements Serialize?

Reply:
I see a 👍🏻 on this but not a corresponding change in the latest commit.

Reply:
We now capture the time without the serializer:
```
                "time" => FnValue(|_| {
                    chrono::Local::now().to_rfc3339()
                }),
```